buildsystem/coverage_report.gradle

```diff
@@ -42,24 +90,137 @@ if (variants != null) {
+                dir: project.buildDir,
+                includes: [
+                        'jacoco/testDebugUnitTest.exec',
+                        'outputs/code-coverage/connected/*coverage.ec'
+                ]
+        )
+    }
+}
+
+def setupAndroidCoverageVerification(threshold) {
+    task jacocoTestCoverageVerification(
+            type: JacocoCoverageVerification,
+            dependsOn: [
+                    'testDebugUnitTest',
+                    'testExistenceValidation'
+            ]
+    ) {
+        violationRules {
+            rule {
+                limit {
+                    minimum = threshold
+                }
+            }
+        }
+        final def coverageSourceDirs = [
+                "$projectDir/src/main/java",
+                "$projectDir/src/main/kotlin"
+        ]
+        final def debugTree = fileTree(
+                dir: "$buildDir/tmp/kotlin-classes/debug",
+                excludes: jacocoThresholdFileFilter
+        )
+        sourceDirectories.from = files(coverageSourceDirs)
+        classDirectories.from = files([debugTree])
+        executionData.from = fileTree(
+                dir: project.buildDir,
+                includes: ['jacoco/testDebugUnitTest.exec']
+        )
+    }
+}
+
+def setupKotlinCoverageVerification(threshold) {
+    jacocoTestCoverageVerification {
+        dependsOn testDebugUnitTest
+        dependsOn testExistenceValidation
+
+        violationRules {
+            rule {
+                limit {
+                    minimum = threshold
+                }
             }
+        }
+    }
+}
+
+def setupTestExistenceValidationTask() {
+    task testExistenceValidation(type: TestExistenceValidation)
+}
+
+project.ext {
+    jacocoCoverageThresholdDefault = 0.60
+
+    jacocoIgnoreList = [] // add module names to ignore
+    // Exclude file by names, packages or types. Such files will be ignored during test coverage
+    // calculation
+    jacocoThresholdFileFilter = [
+            '**/*App.*',
+            '**/*Application.*',
+            '**/*Activity.*',
+            '**/*Fragment.*',
+            '**/*View.*',
+            '**/*ViewGroup.*',
+            '**/*JsonAdapter.*',
+            '**/di/**',
+            '**/*Dagger.*',
+            '**/*Binding*.*'
+    ]
+}
 
-            sourceDirectories.from = files([mainSrc])
-            classDirectories.from = files([debugTree], [kotlinTree])
-            executionData.from = fileTree(dir: "$buildDir", includes: [
-                    "jacoco/test${variantName}UnitTest.exec",
-                    "outputs/code-coverage/connected/*coverage.ec"
-            ])
+private static boolean isAndroidModule(Project project) {
+    def isAndroidLibrary = project.plugins.hasPlugin('com.android.library')
+    def isAndroidApp = project.plugins.hasPlugin('com.android.application')
+    return isAndroidLibrary || isAndroidApp
+}
+
+// Modules without any unit tests should also fail validation
```

Comment:
would it fail build if validation fails? We don't have tests for :contracts module (an android module)

Reply:
If its not included in the modules build.gradle, then it doesn't fail.